### PR TITLE
Wrapper for byte array

### PR DESCRIPTION
**Pull request by TharinduDG**
[ISSUE-178](https://github.com/ScorexFoundation/Scorex/issues/178)

### CODE REVIEW

**Comment by catena2w** [comment]
Please:
1. Resolve conflicts
2. Define equals
3. Define hashCode
4. Define toString

**Comment by TharinduDG** [comment]
@catena2w Done.
I had to use `!@@` in some places. Please check if its ok or not.

**Comment by kushti** [comment]
@TharinduDG please resolve the conflicts!

**Comment by TharinduDG** [comment]
@kushti Done. Please check.

File: examples/src/main/scala/examples/hybrid/validation/ParentBlockValidator.scala [hunk 1]
```diff
@@ -14,16 +15,16 @@ class ParentBlockValidator(storage: HistoryStorage)
...
       case powBlock: PowBlock => if (!storage.isGenesis(powBlock)) {
         //check PoW parent id ???
-        require(storage.modifierById(powBlock.parentId).isDefined, s"Parent ${Base58.encode(powBlock.parentId)} missed")
+        require(storage.modifierById(ModifierId !@@ powBlock.parentId.arr).isDefined, s"Parent ${Base58.encode(powBlock.parentId.arr)} missed")
         //check referenced PoS block exists as well
         val posBlock = storage.modifierById(powBlock.prevPosId).get
 
         //check referenced PoS block points to parent PoW block
-        require(posBlock.parentId sameElements posBlock.parentId, "ref rule broken")
+        require(posBlock.parentId.arr sameElements posBlock.parentId.arr, "ref rule broken")
```
Comment (kushti) [inline examples/src/main/scala/examples/hybrid/validation/ParentBlockValidator.scala]: should be "posBlock.parentId == posBlock.parentId" , as a particular motivation to do boxing was to have "==" operator working (please check all other sameElements occurences, there are a lot of them!)

**Comment by kushti** [comment]
@ceilican arrays (of primitive types) are about much less memory footprint, see http://www.lihaoyi.com/post/BenchmarkingScalaCollections.html#arrays-are-great . maybe WrappedArray is the solution

**Comment by ceilican** [comment]
@kushti , `WrappedArray` is a `Seq`. So, your suggestion is a special case of my suggestion.

Note that we don't need to use `WrappedArray` directly. Whenever we have an `Array` and a `Seq` is expected, an implicit conversion from `Array` to `WrappedArray` will occur.  (Of course, though, we might want to use `WrappedArray` directly to minimize the number of times that such implicit conversions occur.)

My main point is: I think we should use `Seq` instead of `Array`, because `Array` is mutable, bug-prone, java-ish, ...  Which type of `Seq` (e.g. `WrappedArray`, `List`, `Vector`,...) doesn't matter so much to me.

So, what if we do the following:
1. When we receive raw data from another peer, we create a `WrappedArray` of bytes, instead of an `Array` of bytes. (as you suggested.)
2. Everywhere else in the code, where the data is consumed, we use the more general `Seq` (of which `WrappedArray` is just a particular subclass).

This would mean that, everywhere else in the code, we are not committing ourselves to any particular implementation of `Seq`. In the future, we could even experiment changing `WrappedArray` to `Vector` or even to our own optimized implementation of `Seq`, if we wish.

**Comment by ceilican** [comment]
@kushti, what do you think?

**Comment by kushti** [comment]
@ceilican while WrappedArray is implementing Seq, under the hood it is just a wrapper for an Array, see Wrapped.ofByte class definition. 

Around the project, we're using ByteStr, ByteString(from Akka), ByteBuffer. There's also ByteArrayWrapper in IODB (which is basically WrappedArray, but with faster comparison). I'm for any option (WrappedArray or maybe ByteString), but lets' be consistent around the code.

**Comment by ceilican** [comment]
@kushti 

> @ceilican while WrappedArray is implementing Seq, under the hood it is just a wrapper for an Array, see Wrapped.ofByte class definition.

> Around the project, we're using ByteStr, ByteString(from Akka), ByteBuffer. There's also ByteArrayWrapper in IODB (which is basically WrappedArray, but with faster comparison). I'm for any option (WrappedArray or maybe ByteString), but lets' be consistent around the code.

Ok. I will write an issue about this. 

What should we do this PR then? I think it will become obsolete once we replace all `Array[Byte]` by, let's say,`ByteString`.  Should we close it?

**Comment by kushti** [comment]
@ceilican yes, I think this it is better to close this PR

**Pull request closed without merging**